{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d643d56c-9eb1-49d5-85f2-cc92e5586d2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d40016c3-750d-42b9-8d95-809b0f3b5f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = \"https://docs.google.com/spreadsheets/d/18_iU1P2Oe1U5jKgYJ1htMb71PZCEExnkochUIphEByQ/edit?resourcekey=&gid=1314325290#gid=1314325290\"\n",
    "sheet_id = url.split('/')[-2]\n",
    "gid = url.split(\"gid=\")[-1]\n",
    "csv_url = f\"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=csv&gid={gid}\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6a7b421e-097e-4ce5-9f0a-f9004dfa2f26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Timestamp</th>\n",
       "      <th>Which is the most important thing to cover on day 10?</th>\n",
       "      <th>Which is the second most important thing to cover on day 10?</th>\n",
       "      <th>Which is the third most important thing to cover on day 10?</th>\n",
       "      <th>Which is the fourth most important thing to cover on day 10?</th>\n",
       "      <th>Which is the fifth most important thing to cover on day 10?</th>\n",
       "      <th>Column 7</th>\n",
       "      <th>Column 8</th>\n",
       "      <th>Column 9</th>\n",
       "      <th>Column 10</th>\n",
       "      <th>Column 11</th>\n",
       "      <th>Column 12</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7/17/2025 11:10:17</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7/17/2025 11:14:31</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Wish we got more experience with writing our o...</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>Adapting DESeq for T21 data - when should you ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7/17/2025 11:15:55</td>\n",
       "      <td>making my own git hub repository</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7/17/2025 11:16:39</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7/17/2025 11:17:44</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Exploring a new supercomputer (picking the rig...</td>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7/17/2025 11:20:06</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>Some more information on how to make ggplot ma...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>making my own git hub repository</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7/17/2025 11:20:38</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7/17/2025 11:20:47</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7/17/2025 11:21:03</td>\n",
       "      <td>I have come across ‘peak recalls’ in ModENCODE...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>7/17/2025 11:21:11</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>7/17/2025 11:21:20</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>7/17/2025 11:22:29</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7/17/2025 11:22:59</td>\n",
       "      <td>Wish we got more experience with writing our o...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>Some more information on how to make ggplot ma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>7/17/2025 11:23:12</td>\n",
       "      <td>Sequencing techniques to find histone demethyl...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Exploring a new supercomputer (picking the rig...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>7/17/2025 11:23:40</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>making my own git hub repository</td>\n",
       "      <td>Adapting DESeq for T21 data - when should you ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>7/17/2025 11:23:42</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Exploring a new supercomputer (picking the rig...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>7/17/2025 11:24:05</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>7/17/2025 11:25:28</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>7/17/2025 11:26:40</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>7/17/2025 11:32:21</td>\n",
       "      <td>Adapting DESeq for T21 data - when should you ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Some more information on how to make ggplot ma...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>7/17/2025 11:38:02</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>making my own git hub repository</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>7/17/2025 11:40:57</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>Sequencing techniques to find histone demethyl...</td>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>7/17/2025 11:42:11</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>Adapting DESeq for T21 data - when should you ...</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Timestamp Which is the most important thing to cover on day 10?  \\\n",
       "0   7/17/2025 11:10:17  More about how to interpret different results ...      \n",
       "1   7/17/2025 11:14:31  More about how to interpret different results ...      \n",
       "2   7/17/2025 11:15:55                   making my own git hub repository      \n",
       "3   7/17/2025 11:16:39  scRNAseq library prep, qc steps and analysis, ...      \n",
       "4   7/17/2025 11:17:44  More about how to interpret different results ...      \n",
       "5   7/17/2025 11:20:06      Now I have a Gene list, what do I do with it?      \n",
       "6   7/17/2025 11:20:38  Comparison of different tools for some of the ...      \n",
       "7   7/17/2025 11:20:47  Let’s learn AWK (all that promised fun!!!!) an...      \n",
       "8   7/17/2025 11:21:03  I have come across ‘peak recalls’ in ModENCODE...      \n",
       "9   7/17/2025 11:21:11  Some of the statistics for RNA-Seq - i.e., aft...      \n",
       "10  7/17/2025 11:21:20  Let’s learn AWK (all that promised fun!!!!) an...      \n",
       "11  7/17/2025 11:22:29  Databases that contain public RNA-seq (and oth...      \n",
       "12  7/17/2025 11:22:59  Wish we got more experience with writing our o...      \n",
       "13  7/17/2025 11:23:12  Sequencing techniques to find histone demethyl...      \n",
       "14  7/17/2025 11:23:40  Some of the statistics for RNA-Seq - i.e., aft...      \n",
       "15  7/17/2025 11:23:42  More about how to interpret different results ...      \n",
       "16  7/17/2025 11:24:05  Databases that contain public RNA-seq (and oth...      \n",
       "17  7/17/2025 11:25:28  Some of the statistics for RNA-Seq - i.e., aft...      \n",
       "18  7/17/2025 11:26:40      Now I have a Gene list, what do I do with it?      \n",
       "19  7/17/2025 11:32:21  Adapting DESeq for T21 data - when should you ...      \n",
       "20  7/17/2025 11:38:02  Databases that contain public RNA-seq (and oth...      \n",
       "21  7/17/2025 11:40:57  scRNAseq library prep, qc steps and analysis, ...      \n",
       "22  7/17/2025 11:42:11  Databases that contain public RNA-seq (and oth...      \n",
       "\n",
       "   Which is the second most important thing to cover on day 10?  \\\n",
       "0   Some of the statistics for RNA-Seq - i.e., aft...             \n",
       "1   Some of the statistics for RNA-Seq - i.e., aft...             \n",
       "2   More about how to interpret different results ...             \n",
       "3   Databases that contain public RNA-seq (and oth...             \n",
       "4   Exploring a new supercomputer (picking the rig...             \n",
       "5   Pro tips about choosing reference and annotati...             \n",
       "6   If there are low-cost options for super comput...             \n",
       "7   scRNAseq library prep, qc steps and analysis, ...             \n",
       "8   More about how to interpret different results ...             \n",
       "9   Comparison of different tools for some of the ...             \n",
       "10  Comparison of different tools for some of the ...             \n",
       "11  More about how to interpret different results ...             \n",
       "12  More about how to interpret different results ...             \n",
       "13  Databases that contain public RNA-seq (and oth...             \n",
       "14  scRNAseq library prep, qc steps and analysis, ...             \n",
       "15  Some of the statistics for RNA-Seq - i.e., aft...             \n",
       "16  More about how to interpret different results ...             \n",
       "17  More about how to interpret different results ...             \n",
       "18  Pro tips about choosing reference and annotati...             \n",
       "19  Some of the statistics for RNA-Seq - i.e., aft...             \n",
       "20  If there are low-cost options for super comput...             \n",
       "21      Now I have a Gene list, what do I do with it?             \n",
       "22                                                NaN             \n",
       "\n",
       "   Which is the third most important thing to cover on day 10?  \\\n",
       "0   Pro tips about choosing reference and annotati...            \n",
       "1   Wish we got more experience with writing our o...            \n",
       "2   Let’s learn AWK (all that promised fun!!!!) an...            \n",
       "3   Pro tips about choosing reference and annotati...            \n",
       "4     DESeq vs different DEG calling software/methods            \n",
       "5   Some more information on how to make ggplot ma...            \n",
       "6   If there are low-cost options for super comput...            \n",
       "7                        Variant calling in scRNA seq            \n",
       "8   Some of the statistics for RNA-Seq - i.e., aft...            \n",
       "9       Now I have a Gene list, what do I do with it?            \n",
       "10  Databases that contain public RNA-seq (and oth...            \n",
       "11  Some of the statistics for RNA-Seq - i.e., aft...            \n",
       "12  Some of the statistics for RNA-Seq - i.e., aft...            \n",
       "13  Pro tips about choosing reference and annotati...            \n",
       "14  Comparison of different tools for some of the ...            \n",
       "15  Databases that contain public RNA-seq (and oth...            \n",
       "16  Some of the statistics for RNA-Seq - i.e., aft...            \n",
       "17                       Variant calling in scRNA seq            \n",
       "18                                                NaN            \n",
       "19    DESeq vs different DEG calling software/methods            \n",
       "20                   making my own git hub repository            \n",
       "21  Databases that contain public RNA-seq (and oth...            \n",
       "22  Virtual environments for python and R so I don...            \n",
       "\n",
       "   Which is the fourth most important thing to cover on day 10?  \\\n",
       "0   Databases that contain public RNA-seq (and oth...             \n",
       "1   Virtual environments for python and R so I don...             \n",
       "2   Databases that contain public RNA-seq (and oth...             \n",
       "3   If there are low-cost options for super comput...             \n",
       "4   Virtual environments for python and R so I don...             \n",
       "5   Comparison of different tools for some of the ...             \n",
       "6   scRNAseq library prep, qc steps and analysis, ...             \n",
       "7   Some of the statistics for RNA-Seq - i.e., aft...             \n",
       "8       Now I have a Gene list, what do I do with it?             \n",
       "9   Virtual environments for python and R so I don...             \n",
       "10  Pro tips about choosing reference and annotati...             \n",
       "11  Comparison of different tools for some of the ...             \n",
       "12  Comparison of different tools for some of the ...             \n",
       "13  scRNAseq library prep, qc steps and analysis, ...             \n",
       "14                   making my own git hub repository             \n",
       "15      Now I have a Gene list, what do I do with it?             \n",
       "16      Now I have a Gene list, what do I do with it?             \n",
       "17      Now I have a Gene list, what do I do with it?             \n",
       "18  Databases that contain public RNA-seq (and oth...             \n",
       "19  scRNAseq library prep, qc steps and analysis, ...             \n",
       "20  More about how to interpret different results ...             \n",
       "21  Sequencing techniques to find histone demethyl...             \n",
       "22  Adapting DESeq for T21 data - when should you ...             \n",
       "\n",
       "   Which is the fifth most important thing to cover on day 10?  Column 7  \\\n",
       "0   Comparison of different tools for some of the ...                NaN   \n",
       "1   Adapting DESeq for T21 data - when should you ...                NaN   \n",
       "2   Some of the statistics for RNA-Seq - i.e., aft...                NaN   \n",
       "3     DESeq vs different DEG calling software/methods                NaN   \n",
       "4   Pro tips about choosing reference and annotati...                NaN   \n",
       "5                    making my own git hub repository                NaN   \n",
       "6     DESeq vs different DEG calling software/methods                NaN   \n",
       "7   If there are low-cost options for super comput...                NaN   \n",
       "8     DESeq vs different DEG calling software/methods                NaN   \n",
       "9   More about how to interpret different results ...                NaN   \n",
       "10  scRNAseq library prep, qc steps and analysis, ...                NaN   \n",
       "11  Pro tips about choosing reference and annotati...                NaN   \n",
       "12  Some more information on how to make ggplot ma...                NaN   \n",
       "13  Exploring a new supercomputer (picking the rig...                NaN   \n",
       "14  Adapting DESeq for T21 data - when should you ...                NaN   \n",
       "15  Exploring a new supercomputer (picking the rig...                NaN   \n",
       "16  If there are low-cost options for super comput...                NaN   \n",
       "17  Virtual environments for python and R so I don...                NaN   \n",
       "18  Let’s learn AWK (all that promised fun!!!!) an...                NaN   \n",
       "19  Some more information on how to make ggplot ma...                NaN   \n",
       "20  Some of the statistics for RNA-Seq - i.e., aft...                NaN   \n",
       "21                       Variant calling in scRNA seq                NaN   \n",
       "22      Now I have a Gene list, what do I do with it?                NaN   \n",
       "\n",
       "    Column 8  Column 9  Column 10  Column 11  Column 12  \n",
       "0        NaN       NaN        NaN        NaN        NaN  \n",
       "1        NaN       NaN        NaN        NaN        NaN  \n",
       "2        NaN       NaN        NaN        NaN        NaN  \n",
       "3        NaN       NaN        NaN        NaN        NaN  \n",
       "4        NaN       NaN        NaN        NaN        NaN  \n",
       "5        NaN       NaN        NaN        NaN        NaN  \n",
       "6        NaN       NaN        NaN        NaN        NaN  \n",
       "7        NaN       NaN        NaN        NaN        NaN  \n",
       "8        NaN       NaN        NaN        NaN        NaN  \n",
       "9        NaN       NaN        NaN        NaN        NaN  \n",
       "10       NaN       NaN        NaN        NaN        NaN  \n",
       "11       NaN       NaN        NaN        NaN        NaN  \n",
       "12       NaN       NaN        NaN        NaN        NaN  \n",
       "13       NaN       NaN        NaN        NaN        NaN  \n",
       "14       NaN       NaN        NaN        NaN        NaN  \n",
       "15       NaN       NaN        NaN        NaN        NaN  \n",
       "16       NaN       NaN        NaN        NaN        NaN  \n",
       "17       NaN       NaN        NaN        NaN        NaN  \n",
       "18       NaN       NaN        NaN        NaN        NaN  \n",
       "19       NaN       NaN        NaN        NaN        NaN  \n",
       "20       NaN       NaN        NaN        NaN        NaN  \n",
       "21       NaN       NaN        NaN        NaN        NaN  \n",
       "22       NaN       NaN        NaN        NaN        NaN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df = pd.read_csv(csv_url)\n",
    "df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7774a504-7d05-4bcc-b35b-daafb1b28c17",
   "metadata": {},
   "outputs": [],
   "source": [
    "countcols = ['Which is the most important thing to cover on day 10?',\n",
    "       'Which is the second most important thing to cover on day 10?',\n",
    "       'Which is the third most important thing to cover on day 10?',\n",
    "       'Which is the fourth most important thing to cover on day 10?',\n",
    "       'Which is the fifth most important thing to cover on day 10?']\n",
    "\n",
    "points = [5,4,3,2,1]\n",
    "names = [\"1st\", \"2nd\", \"3rd\", \"4th\", \"5th\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ce4290dc-b684-43e0-abd7-80899e043f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "vote_dfs = []\n",
    "for i, colname in enumerate(countcols):\n",
    "    votes = df[colname].value_counts().to_frame()\n",
    "    votes[\"questionrank\"] = names[i]\n",
    "    votes[\"point_increase\"] = points[i]\n",
    "    votes[\"q\"] = votes.index\n",
    "    vote_dfs.append(votes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "33a28565-4573-4875-90a2-dd7aa702519c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "result = pd.concat(vote_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "29b11f42-b31b-4ff8-b845-989062445c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>questionrank</th>\n",
       "      <th>point_increase</th>\n",
       "      <th>q</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>More about how to interpret different results and QC data. Now we know how to get the output files, but what to do in terms of interpreting the results would be helpful.</th>\n",
       "      <td>4</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Databases that contain public RNA-seq (and other omics) data</th>\n",
       "      <td>4</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Some of the statistics for RNA-Seq - i.e., after you have your results, what do you do to get to publication</th>\n",
       "      <td>3</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scRNAseq library prep, qc steps and analysis, scRNA seq graphing</th>\n",
       "      <td>2</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Now I have a Gene list, what do I do with it?</th>\n",
       "      <td>2</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scRNAseq library prep, qc steps and analysis, scRNA seq graphing</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Virtual environments for python and R so I don't get stuck in install hell</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Let’s learn AWK (all that promised fun!!!!) and SED, please.</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant calling in scRNA seq</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Now I have a Gene list, what do I do with it?</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    count questionrank  \\\n",
       "More about how to interpret different results a...      4          1st   \n",
       "Databases that contain public RNA-seq (and othe...      4          1st   \n",
       "Some of the statistics for RNA-Seq - i.e., afte...      3          1st   \n",
       "scRNAseq library prep, qc steps and analysis, s...      2          1st   \n",
       "Now I have a Gene list, what do I do with it?           2          1st   \n",
       "...                                                   ...          ...   \n",
       "scRNAseq library prep, qc steps and analysis, s...      1          5th   \n",
       "Virtual environments for python and R so I don'...      1          5th   \n",
       "Let’s learn AWK (all that promised fun!!!!) and...      1          5th   \n",
       "Variant calling in scRNA seq                            1          5th   \n",
       "Now I have a Gene list, what do I do with it?           1          5th   \n",
       "\n",
       "                                                    point_increase  \\\n",
       "More about how to interpret different results a...               5   \n",
       "Databases that contain public RNA-seq (and othe...               5   \n",
       "Some of the statistics for RNA-Seq - i.e., afte...               5   \n",
       "scRNAseq library prep, qc steps and analysis, s...               5   \n",
       "Now I have a Gene list, what do I do with it?                    5   \n",
       "...                                                            ...   \n",
       "scRNAseq library prep, qc steps and analysis, s...               1   \n",
       "Virtual environments for python and R so I don'...               1   \n",
       "Let’s learn AWK (all that promised fun!!!!) and...               1   \n",
       "Variant calling in scRNA seq                                     1   \n",
       "Now I have a Gene list, what do I do with it?                    1   \n",
       "\n",
       "                                                                                                    q  \n",
       "More about how to interpret different results a...  More about how to interpret different results ...  \n",
       "Databases that contain public RNA-seq (and othe...  Databases that contain public RNA-seq (and oth...  \n",
       "Some of the statistics for RNA-Seq - i.e., afte...  Some of the statistics for RNA-Seq - i.e., aft...  \n",
       "scRNAseq library prep, qc steps and analysis, s...  scRNAseq library prep, qc steps and analysis, ...  \n",
       "Now I have a Gene list, what do I do with it?           Now I have a Gene list, what do I do with it?  \n",
       "...                                                                                               ...  \n",
       "scRNAseq library prep, qc steps and analysis, s...  scRNAseq library prep, qc steps and analysis, ...  \n",
       "Virtual environments for python and R so I don'...  Virtual environments for python and R so I don...  \n",
       "Let’s learn AWK (all that promised fun!!!!) and...  Let’s learn AWK (all that promised fun!!!!) an...  \n",
       "Variant calling in scRNA seq                                             Variant calling in scRNA seq  \n",
       "Now I have a Gene list, what do I do with it?           Now I have a Gene list, what do I do with it?  \n",
       "\n",
       "[61 rows x 4 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "bb483ad9-96b7-4786-a5fb-96bbdcfc6c40",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>questionrank</th>\n",
       "      <th>point_increase</th>\n",
       "      <th>q</th>\n",
       "      <th>total_points_per_Q</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>More about how to interpret different results and QC data. Now we know how to get the output files, but what to do in terms of interpreting the results would be helpful.</th>\n",
       "      <td>4</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Databases that contain public RNA-seq (and other omics) data</th>\n",
       "      <td>4</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Some of the statistics for RNA-Seq - i.e., after you have your results, what do you do to get to publication</th>\n",
       "      <td>3</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scRNAseq library prep, qc steps and analysis, scRNA seq graphing</th>\n",
       "      <td>2</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Now I have a Gene list, what do I do with it?</th>\n",
       "      <td>2</td>\n",
       "      <td>1st</td>\n",
       "      <td>5</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>scRNAseq library prep, qc steps and analysis, scRNA seq graphing</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Virtual environments for python and R so I don't get stuck in install hell</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Let’s learn AWK (all that promised fun!!!!) and SED, please.</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Variant calling in scRNA seq</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Now I have a Gene list, what do I do with it?</th>\n",
       "      <td>1</td>\n",
       "      <td>5th</td>\n",
       "      <td>1</td>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    count questionrank  \\\n",
       "More about how to interpret different results a...      4          1st   \n",
       "Databases that contain public RNA-seq (and othe...      4          1st   \n",
       "Some of the statistics for RNA-Seq - i.e., afte...      3          1st   \n",
       "scRNAseq library prep, qc steps and analysis, s...      2          1st   \n",
       "Now I have a Gene list, what do I do with it?           2          1st   \n",
       "...                                                   ...          ...   \n",
       "scRNAseq library prep, qc steps and analysis, s...      1          5th   \n",
       "Virtual environments for python and R so I don'...      1          5th   \n",
       "Let’s learn AWK (all that promised fun!!!!) and...      1          5th   \n",
       "Variant calling in scRNA seq                            1          5th   \n",
       "Now I have a Gene list, what do I do with it?           1          5th   \n",
       "\n",
       "                                                    point_increase  \\\n",
       "More about how to interpret different results a...               5   \n",
       "Databases that contain public RNA-seq (and othe...               5   \n",
       "Some of the statistics for RNA-Seq - i.e., afte...               5   \n",
       "scRNAseq library prep, qc steps and analysis, s...               5   \n",
       "Now I have a Gene list, what do I do with it?                    5   \n",
       "...                                                            ...   \n",
       "scRNAseq library prep, qc steps and analysis, s...               1   \n",
       "Virtual environments for python and R so I don'...               1   \n",
       "Let’s learn AWK (all that promised fun!!!!) and...               1   \n",
       "Variant calling in scRNA seq                                     1   \n",
       "Now I have a Gene list, what do I do with it?                    1   \n",
       "\n",
       "                                                                                                    q  \\\n",
       "More about how to interpret different results a...  More about how to interpret different results ...   \n",
       "Databases that contain public RNA-seq (and othe...  Databases that contain public RNA-seq (and oth...   \n",
       "Some of the statistics for RNA-Seq - i.e., afte...  Some of the statistics for RNA-Seq - i.e., aft...   \n",
       "scRNAseq library prep, qc steps and analysis, s...  scRNAseq library prep, qc steps and analysis, ...   \n",
       "Now I have a Gene list, what do I do with it?           Now I have a Gene list, what do I do with it?   \n",
       "...                                                                                               ...   \n",
       "scRNAseq library prep, qc steps and analysis, s...  scRNAseq library prep, qc steps and analysis, ...   \n",
       "Virtual environments for python and R so I don'...  Virtual environments for python and R so I don...   \n",
       "Let’s learn AWK (all that promised fun!!!!) and...  Let’s learn AWK (all that promised fun!!!!) an...   \n",
       "Variant calling in scRNA seq                                             Variant calling in scRNA seq   \n",
       "Now I have a Gene list, what do I do with it?           Now I have a Gene list, what do I do with it?   \n",
       "\n",
       "                                                    total_points_per_Q  \n",
       "More about how to interpret different results a...                  20  \n",
       "Databases that contain public RNA-seq (and othe...                  20  \n",
       "Some of the statistics for RNA-Seq - i.e., afte...                  15  \n",
       "scRNAseq library prep, qc steps and analysis, s...                  10  \n",
       "Now I have a Gene list, what do I do with it?                       10  \n",
       "...                                                                ...  \n",
       "scRNAseq library prep, qc steps and analysis, s...                   1  \n",
       "Virtual environments for python and R so I don'...                   1  \n",
       "Let’s learn AWK (all that promised fun!!!!) and...                   1  \n",
       "Variant calling in scRNA seq                                         1  \n",
       "Now I have a Gene list, what do I do with it?                        1  \n",
       "\n",
       "[61 rows x 5 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[\"total_points_per_Q\"] = result[\"count\"]*result[\"point_increase\"]\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "24b04198-f264-46b3-a24e-c8e5987309e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>q</th>\n",
       "      <th>total_points_per_Q</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>More about how to interpret different results ...</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Some of the statistics for RNA-Seq - i.e., aft...</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Databases that contain public RNA-seq (and oth...</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Now I have a Gene list, what do I do with it?</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>scRNAseq library prep, qc steps and analysis, ...</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Comparison of different tools for some of the ...</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Pro tips about choosing reference and annotati...</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>If there are low-cost options for super comput...</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Let’s learn AWK (all that promised fun!!!!) an...</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>making my own git hub repository</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Virtual environments for python and R so I don...</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DESeq vs different DEG calling software/methods</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adapting DESeq for T21 data - when should you ...</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Wish we got more experience with writing our o...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Sequencing techniques to find histone demethyl...</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Variant calling in scRNA seq</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Exploring a new supercomputer (picking the rig...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I have come across ‘peak recalls’ in ModENCODE...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Some more information on how to make ggplot ma...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    q  total_points_per_Q\n",
       "8   More about how to interpret different results ...                  47\n",
       "13  Some of the statistics for RNA-Seq - i.e., aft...                  47\n",
       "3   Databases that contain public RNA-seq (and oth...                  43\n",
       "9       Now I have a Gene list, what do I do with it?                  26\n",
       "18  scRNAseq library prep, qc steps and analysis, ...                  25\n",
       "1   Comparison of different tools for some of the ...                  23\n",
       "10  Pro tips about choosing reference and annotati...                  21\n",
       "6   If there are low-cost options for super comput...                  15\n",
       "7   Let’s learn AWK (all that promised fun!!!!) an...                  14\n",
       "17                   making my own git hub repository                  11\n",
       "15  Virtual environments for python and R so I don...                  10\n",
       "2     DESeq vs different DEG calling software/methods                   9\n",
       "0   Adapting DESeq for T21 data - when should you ...                   9\n",
       "16  Wish we got more experience with writing our o...                   8\n",
       "11  Sequencing techniques to find histone demethyl...                   7\n",
       "14                       Variant calling in scRNA seq                   7\n",
       "4   Exploring a new supercomputer (picking the rig...                   6\n",
       "5   I have come across ‘peak recalls’ in ModENCODE...                   5\n",
       "12  Some more information on how to make ggplot ma...                   5"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_sorted = result_grouped.sort_values(by=\"total_points_per_Q\", ascending=False)\n",
    "result_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f873e9b-5e23-4039-9dcb-6bc250913f3d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
